source/simulator/melt.cc

```diff
@@ -0,0 +1,1092 @@
+
+      Assert (material_model_outputs.densities[q_point] > 0,
+              ExcMessage ("The density needs to be a positive quantity "
+                          "when melt transport is included in the simulation."));
+
+      const double melting_rate         = material_model_outputs.reaction_terms[q_point][advection_field.compositional_variable];
+      const double density              = material_model_outputs.densities[q_point];
+      const double current_phi          = material_model_inputs.composition[q_point][advection_field.compositional_variable];
+      const double divergence_u         = scratch.current_velocity_divergences[q_point];
+      const double compressibility      = (this->get_material_model().is_compressible()
+                                           ?
+                                           material_model_outputs.compressibilities[q_point]
+                                           :
+                                           0.0);
+      const Tensor<1,dim> current_u     = scratch.current_velocity_values[q_point];
+      const Tensor<1,dim>
+      gravity = this->get_gravity_model().gravity_vector (scratch.finite_element_values.quadrature_point(q_point));
+
+      double melt_transport_RHS = melting_rate / density
+                                  + divergence_u + compressibility * density * (current_u * gravity);
+
+      if (current_phi < this->get_melt_handler().melt_transport_threshold
+          && melting_rate < this->get_melt_handler().melt_transport_threshold)
+        melt_transport_RHS = melting_rate / density;
+
+      return melt_transport_RHS;
+    }
+
+
+    namespace OtherTerms
+    {
+      template <int dim>
+      void
+      pressure_rhs_compatibility_modification_melt (const SimulatorAccess<dim>                      &simulator_access,
+                                                    internal::Assembly::Scratch::StokesSystem<dim>  &scratch,
+                                                    internal::Assembly::CopyData::StokesSystem<dim> &data)
+      {
+        const Introspection<dim> &introspection = simulator_access.introspection();
+
+        const unsigned int dofs_per_cell = scratch.finite_element_values.get_fe().dofs_per_cell;
+        const unsigned int n_q_points    = scratch.finite_element_values.n_quadrature_points;
+
+        const FEValuesExtractors::Scalar ex_p_f = introspection.variable("fluid pressure").extractor_scalar();
+
+        for (unsigned int q=0; q<n_q_points; ++q)
+          for (unsigned int i=0; i<dofs_per_cell; ++i)
+            {
+              scratch.phi_p[i] = scratch.finite_element_values[ex_p_f].value (i, q);
+              data.local_pressure_shape_function_integrals(i) += scratch.phi_p[i] * scratch.finite_element_values.JxW(q);
+            }
+      }
+    }
+  }
+
+
+  template <int dim>
+  void
+  MeltHandler<dim>::
+  compute_melt_variables(LinearAlgebra::BlockVector &solution)
+  {
+    if (!this->include_melt_transport())
+      return;
```

Comment:
I would almost prefer if you would just put this into an assertion. You already check in all of the places where you call the function that a melt handler exists, so we should not be calling this function if melt transport is not enabled -- let's just make sure that's the case.